src/skadi/sample.cc

```diff
@@ -110,47 +111,83 @@ class cache_item_t {
     return usages;
   }
 
-  inline const char* get_unpacked() {
+  inline char* get_unpacked() {
     return unpacked;
   }
 
-  inline const char* detach_unpacked() {
+  inline char* detach_unpacked() {
     auto rv = unpacked;
     unpacked = nullptr;
     return rv;
   }
 
-  bool unpack(const char* unpacked) {
+  bool unpack(char* unpacked) {
     this->unpacked = unpacked;
-    // for setting where to read compressed data from
-    auto src_func = [this](z_stream& s) -> void {
-      s.next_in = static_cast<Byte*>(static_cast<void*>(data.get()));
-      s.avail_in = static_cast<unsigned int>(data.size());
-    };
-
-    // for setting where to write the uncompressed data to
-    auto dst_func = [this](z_stream& s) -> int {
-      s.next_out = (Byte*)(this->unpacked);
-      s.avail_out = HGT_BYTES;
-      return Z_FINISH; // we know the output will hold all the input
-    };
-
-    // we have to unzip it
-    if (!baldr::inflate(src_func, dst_func)) {
-      LOG_WARN("Corrupt compressed elevation data");
+
+    if (format == format_t::GZIP) {
+      // for setting where to read compressed data from
+      auto src_func = [this](z_stream& s) -> void {
+        s.next_in = static_cast<Byte*>(static_cast<void*>(data.get()));
+        s.avail_in = static_cast<unsigned int>(data.size());
+      };
+
+      // for setting where to write the uncompressed data to
+      auto dst_func = [this](z_stream& s) -> int {
+        s.next_out = (Byte*)(this->unpacked);
+        s.avail_out = HGT_BYTES;
+        return Z_FINISH; // we know the output will hold all the input
+      };
+
+      // we have to unzip it
+      if (!baldr::inflate(src_func, dst_func)) {
+        LOG_WARN("Corrupt gzip elevation data");
+        format = format_t::UNKNOWN;
+        return false;
+      }
+    } else if (format == format_t::LZ4) {
+      LZ4F_decompressionContext_t decode;
+      LZ4F_decompressOptions_t options;
+      LZ4F_createDecompressionContext(&decode, LZ4F_VERSION);
+
+      // Take these two values locally, since LZ4F_decompress expects pointers...
+      size_t src_size = data.size();
+      size_t dest_size = HGT_BYTES;
+      size_t result;
+      do {
+        result = LZ4F_decompress(decode, this->unpacked, &dest_size, data.get(), &src_size, &options);
```

Comment:
we can revert all the changes you made to remove `const`ness from this and just cast the `const` away at the time when we need to do that. this is ok because when we create `unpacked` we do so without `const`.

```suggestion
        result = LZ4F_decompress(decode, const_cast<char *>(this->unpacked), &dest_size, data.get(), &src_size, &options);
```